nupic/research/spatial_pooler.py

```diff
@@ -50,8 +51,8 @@ class SpatialPooler(object):
   """
 
   def __init__(self,
+               layer=Layer(columnDimensions=(64,64)),
```

Comment:
This is a dangerous convention.  You should only use immutable types as defaults in function definitions.  Here's a concrete example to demonstrate why:

```python
>>> class MutableObject(object):
...   def __init__(self, foo=None):
...     self.foo = foo
... 
>>> class Adapter(object):
...   def __init__(self, mutableObj=MutableObject(1)):
...     self.mutableObj = mutableObj
... 
>>> obj1 = Adapter()
>>> obj2 = Adapter()
>>> 
>>> obj1.mutableObj.foo = "Foo"
>>> print obj2.mutableObj.foo
Foo
>>> obj2.mutableObj.foo = "Bar"
>>> print obj1.mutableObj.foo
Bar
```

In other words, `obj1.mutableObj` and `obj2.mutableObj` are the same instance of `MutableObject`, despite being attributes of different instances of `Adapter`.  This is because the object is created at time of definition, rather than instantiation.

Using my example, better to have something like the following to handle default values:

```python
class Adapter(object):
  def __init__(self, mutableObj=None):
    self.mutableObj = mutableObj or MutableObject(1)
```